### PR TITLE
Replaces deprecated `Response::type()` by `Response::getType()` (#555)

### LINKED ISSUE #555: Response::type() within AppController's beforeRender()
This is a (multiple allowed):

* [ ] bug
* [X] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Application Skeleton Version: CakePHP 3.5.3

### What I did
I was just looking at the [**3.4 migration guide**](https://book.cakephp.org/3.0/en/appendices/3-4-migration-guide.html#request-response-deprecations)...

### What happened
... when I saw that `Response::type()` had been deprecated and should be replaced by `Response::withType()`.

### What I expected to happen
But when I looked into the new app skeleton provided here, the old `type()` method is still used within the `AppController`'s `beforeRender()` (_JSON / XML_ detection).
Should not this be changed ?

A big thanks 👋

PS : I've dug a bit the WWW, but didn't find a thing about that, am I misunderstanding ?

### PR DESCRIPTION
**Pull request by HorlogeSkynet**
I'm back !

So this is a workaround using `RequestHandler::mapAlias()` and `RequestHandler::responseType()`.
The only problem is the variable `$mimeType` may receive either a `string` or an `array`.
This is why the distinction of cases is done.

I've done some tests on my own, it _should_ work.
BUT, it'd be very interesting if some expert eyes could catch a glimpse of this 👀 

Last point : This is "bigger" than before, and if someone here has a better idea, let it known !

++ 👋

EDIT : 'would fix #555.
EDIT_2 : Sorry for the second commit, didn't know that Travis would complain about that...

### CODE REVIEW

File: src/Controller/AppController.php [hunk 1]
```diff
@@ -63,10 +63,18 @@ public function beforeRender(Event $event)
         // Note: These defaults are just to get started quickly with development
         // and should not be used in production. You should instead set "_serialize"
         // in each action as required.
-        if (!array_key_exists('_serialize', $this->viewVars) &&
-            in_array($this->response->type(), ['application/json', 'application/xml'])
-        ) {
-            $this->set('_serialize', true);
+        if (!array_key_exists('_serialize', $this->viewVars)) {
+            $mimeType = $this->RequestHandler->mapAlias(
```
Comment (markstory) [inline src/Controller/AppController.php]: I'm not sure we can rely on `RequestHandler` being loaded. In addition to this approach you can also use `$this->response->getHeaderLine('Content-Type')` to read the content type header. That might result in simpler code that lets you continue to use `in_array()`.

Comment (HorlogeSkynet) [reply src/Controller/AppController.php]: Why couldn't we rely on `RequestHandler` here ?

--> I didn't know about that one, I may simplify this if you're OK with the concept !

Comment (inoas) [reply src/Controller/AppController.php]: https://gist.github.com/inoas/e58900d3e266a8827b988142af490c6f
... or about the same in class `\Cake\Utility\Text` may help

**Comment by ADmad** [comment]
I am a bit confused, what exactly is the purpose of this change?

**Comment by saeideng** [comment]
>deprecated `Response::type()`
I not found deprecated method
can anyone help to me ?
https://github.com/cakephp/cakephp/blob/fcb0b36eb44460cb5bc50ab6a999b9ab2038ab15/src/Http/Response.php#L1042-L1046

**Comment by HorlogeSkynet** [comment]
As specified during the issue (#555), it's what the [CakePHP 3.4's Migration Guide](https://book.cakephp.org/3.0/en/appendices/3-4-migration-guide.html#request-response-deprecations) announced in the past.

But I don't know either why the source documentation does not specified it. Maybe an oversight ?

**Comment by markstory** [comment]
@HorlogeSkynet The migration notes only document the write side of `type()` being deprecated. With the encoding challeges, we might want to consider adding `getType()`.

**Comment by HorlogeSkynet** [comment]
@markstory This is what is effectively, indeed.
Where are we heading then ?
Do I try something with your `getHeaderLine()`, or do we wait for a proper _getter_ ?

**Comment by ADmad** [comment]
This PR highlights the fact that we are probably missing a method on Response class to get the currently set type/mimetype. There's getMimeType() but it doesn't do what I would intuitively expect. It returns the mapped mimetype for given type instead of currently set mimetype.

**Comment by HorlogeSkynet** [comment]
Well, I couldn't resist, the latest version is simple.
Keep me posted @markstory 👍

**Comment by markstory** [comment]
@ADmad There are higher level abstractions in RequestHandler for setting content types. For lower level APIs there is `withType()`, `withCharset()` and `withHeader()`. What is not 'easy' right now is something like `getType()` which would get the current content-type without the charset bits.

File: src/Controller/AppController.php [hunk 2]
```diff
@@ -55,18 +55,21 @@ public function initialize()
...
-        if (!array_key_exists('_serialize', $this->viewVars) &&
-            in_array($this->response->type(), ['application/json', 'application/xml'])
-        ) {
-            $this->set('_serialize', true);
+        // and should not be used in production.
+        // You should instead set "_serialize" in each action as required.
+        if (!array_key_exists('_serialize', $this->viewVars)) {
+            $contentType = $this->response->getHeaderLine('Content-Type');
+            if (strstr($contentType, 'application/json') ||
+                strstr($contentType, 'application/xml')
```
Comment (markstory) [inline src/Controller/AppController.php]: These `strstr()` calls could be `strpos()`. It is a bit faster and uses less memory. While it is likely not relevant here it is a good practice to adopt.

Comment (HorlogeSkynet) [reply src/Controller/AppController.php]: Yeah it's a beautiful confusion, I had planned to user `strpos()` 😢
May I force push ?

Comment (markstory) [reply src/Controller/AppController.php]: Of course. Force push and rebase as much as you need.

**Comment by saeideng** [comment]
@markstory  I wrote some code/test for Implementing getType()
you can see it in new PR

**Comment by ADmad** [comment]
> What is not 'easy' right now is something like getType() which would get the current content-type without the charset bits.

That's what I was referring to :)

File: src/Controller/AppController.php [hunk 3]
```diff
@@ -55,18 +55,21 @@ public function initialize()
...
+     * @param  \Cake\Event\Event $event The beforeRender event.
      * @return \Cake\Http\Response|null|void
      */
     public function beforeRender(Event $event)
     {
         // Note: These defaults are just to get started quickly with development
-        // and should not be used in production. You should instead set "_serialize"
-        // in each action as required.
-        if (!array_key_exists('_serialize', $this->viewVars) &&
-            in_array($this->response->type(), ['application/json', 'application/xml'])
```
Comment (saeideng) [inline src/Controller/AppController.php]: new Cakephp released 
you can replace this by `$this->response->getType()` now

Comment (HorlogeSkynet) [reply src/Controller/AppController.php]: Yeah thank you, I was waiting for the tag. I start working on this ASAP 👌

File: src/Controller/AppController.php [hunk 4]
```diff
@@ -55,18 +55,21 @@ public function initialize()
     /**
      * Before render callback.
      *
-     * @param \Cake\Event\Event $event The beforeRender event.
+     * @param  \Cake\Event\Event $event The beforeRender event.
```
[the remark this answers is not in the document]
Comment (HorlogeSkynet) [reply src/Controller/AppController.php]: Yeah that was `phpcbf`... Does it matter ?

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Controller/AppController.php]: If you revert it back it won't :)

**Pull request merged**